### PR TITLE
Feature communications struct

### PR DESCRIPTION
**Pull request by guywithhat99**
Overall this PR Created data_packet struct that can be given

- Robot state
- referee Data
- CAN Data
- Sensors

And eventually
- DR16
- logging data

This data can then be packed into a 4kb buffer that can be transmitted to hive to be unpacked by the same struct.

On the hive's end, I'm pretty sure all we have to do is get rid of the pack data function (and do some fun things with the lidar sensors), and everything will be good to go.

On firmware side, everything is good! (I think)

Here is a more comprehensive copilot summary of what I did:

### Sensor Serialization and Inheritance:

* [`src/sensors/Sensor.hpp`](diffhunk://#diff-eabfb106af514607cc3d2e9f44aad2b95f9867d93ab02ad78de9d12e1501fc16R1-R27): Introduced a new `Sensor` base class with `serialize` method and `SensorType` enum.
* [`src/sensors/IMUSensor.hpp`](diffhunk://#diff-084c973c1e48b88712747210cf120adb2eae018b0a296badc9bd769d10953736R8-R25): Updated `IMUSensor` to inherit from the new `Sensor` base class.
* [`src/sensors/ICM20649.hpp`](diffhunk://#diff-413256f20ab021cf3bf757040c358a4aab19fcb0b8013a9899e12428179562adR46-R47): Modified `ICM20649` class to inherit from `Sensor` and implemented the `serialize` method.
* [`src/sensors/TOFSensor.hpp`](diffhunk://#diff-82db4a64faaadf024327891eb1b41a053bfdbd3176bd5422a1e03e37ea0ead19R98-R117): Updated `TOFSensor` to inherit from `Sensor` and added the `serialize` method.
* [`src/sensors/buff_encoder.hpp`](diffhunk://#diff-07298b35ff8be3d85b3f56ff5f3cd3cd489e05bd136b26dfe402fe5f81192639R71-R82): Modified `BuffEncoder` class to inherit from `Sensor` and implemented the `serialize` method.
* [`src/sensors/d200.cpp`](diffhunk://#diff-d87903358acf7e09d59ceb4b5b713befcaab34b00013dbeb44cee671cb8bc14bR199-R208): Updated `D200LD14P` class to inherit from `Sensor` and added the `serialize` method.
*  [`src/comms/data_packet.cpp`](diffhunk://#diff-412075df8be473dcdfe1e4d4226ce556fe3db4577991dea0d60301c4f0149392R1): Included the new `data_packet.hpp` header.

### Code Structure Improvements:

* [`src/constants.hpp`](diffhunk://#diff-6dd9f08c662986128e0efaf41168bd97011c56c9554337eb5fd5151f6bfd1a18R1-R28): Added a new header file to define constants and sensor types.

* [`src/main.cpp`](diffhunk://#diff-34d21af3c614ea3cee120df276c9c4ae95053830d7f1d3deaf009a4625409ad2R147-R173): Included new headers and updated the main loop to handle data packet construction and serialization.

### CODE REVIEW

**Comment by Pandabear1125** [comment]
Make sure you comment the stuff that the doxygen bot is saying is missing

File: src/comms/data_packet.hpp [hunk 1]
```diff
@@ -0,0 +1,512 @@
+#include <cstdint>
```
[the remark this answers is not in the document]
Comment (Pandabear1125) [reply src/comms/data_packet.hpp:1]: Also can you please remove your unused commented code, and comment out the prints in pack/unpackDataPacket

File: src/main.cpp [hunk 2]
```diff
@@ -20,8 +22,8 @@ rm_CAN can;
 RefSystem ref;
 HIDLayer comms;
 
-D200LD14P lidar1(&Serial4, 0);
-D200LD14P lidar2(&Serial5, 1);
+D200LD14P lidar1(&Serial4);
```
Comment (Pandabear1125) [inline src/main.cpp:25]: Why?

File: src/sensors/ICM20649.cpp [hunk 3]
```diff
@@ -2,7 +2,7 @@
 #include <cassert>
 
 // empty constructor
-ICM20649::ICM20649() {}
+// ICM20649::ICM20649() {}
```
Comment (Pandabear1125) [inline src/sensors/ICM20649.cpp:5]: If youre not using this, remove it

File: src/sensors/ICM20649.cpp [hunk 4]
```diff
@@ -99,4 +99,44 @@ void ICM20649::set_gyro_range(int gyro_rate_range) {
...
+void ICM20649::serialize(uint8_t* buffer, size_t& offset) {
+    //returns the acceleration with the offset applied
+    float corrected_accel_X = get_accel_X();
+    float corrected_accel_Y = get_accel_Y();
+    float corrected_accel_Z = get_accel_Z();
+    float corrected_gyro_X = get_gyro_X();
+    float corrected_gyro_Y = get_gyro_Y();
+    float corrected_gyro_Z = get_gyro_Z();
+    //  code that prints out raw bytes of the buffer for this sensor
+    Serial.print("ICM20649: ");
```
Comment (Pandabear1125) [inline src/sensors/ICM20649.cpp:114]: Comment out/remove print

**Review by Pandabear1125 — DISMISSED** [review]
I like how you handle serialization/packing. Great work on that. There is a bit to clean up, but it looks good!

**Review by roccoagain — APPROVED** [review]
LGTM

**Pull request merged**